python/gstaichi/lang/kernel_impl.py

```diff
@@ -927,23 +927,27 @@ def call_back():
                                 "Non contiguous gradient tensors are not supported, please call tensor.grad.contiguous() before passing it into gstaichi kernel."
                             )
 
-                    tmp = v
-                    if (str(v.device) != "cpu") and not (
-                        str(v.device).startswith("cuda") and gstaichi_arch == _ti_core.Arch.cuda
+                    v_taichi = v
+                    grad_taichi = v.grad
+                    if (v.device.type != "cpu") and not (
+                        v.device.type == "cuda" and gstaichi_arch == _ti_core.Arch.cuda
```

Comment:
You're using both `v` and `v_taichi`, which I find confusing. Maybe don't create an alias of `v`?

Reply:
I did this to align with what was originally done. And anyway, creating an alias will be necessary for the gradient.